wpilib/wpilib/pidcontroller.py

```diff
@@ -141,67 +136,76 @@ def __init__(self, Kp, Ki, Kd, *args, **kwargs):
 
         if enabled:
+            feedForward = self.calculateFeedForward()
+
             with self.mutex:
-                input = pidInput.pidGet()
-                self.error = self.getContinuousError(self.setpoint - input)
-
-                if self.pidInput.getPIDSourceType() == self.PIDSourceType.kRate:
-                    if self.P != 0:
-                        self.totalError = self.clamp(
-                            self.totalError + self.error, 
-                            self.minimumOutput / self.P,
-                            self.maximumOutput / self.P)
+                input = self.pidInput.pidGet()
+                pidSourceType = self.pidInput.getPIDSourceType()
+                P = self.P
+                I = self.I
+                D = self.D
+                minimumOutput = self.minimumOutput
+                maximumOutput = self.maximumOutput
+                prevError = self.prevError
+                error = self.getContinuousError(self.setpoint - input)
+                totalError = self.totalError
+
+            # start
+
+            if pidSourceType == self.PIDSourceType.kRate:
+                if P != 0:
+                    totalError = self.clamp(
+                        totalError + error, 
+                        minimumOutput / P,
+                        maximumOutput / P)
                         
-                    self.result = self.P * self.totalError + \
-                        self.D * self.error + \
-                        self.calculateFeedForward()
-
-                else:
-
-                    if self.I != 0:
-                        self.totalError = self.clamp(
-                            self.totalError + self.error,
-                            self.minimumOutput / self.I,
-                            self.maximumOutput / self.I)
+                result = P * totalError + D * error + feedForward
 
-                    self.result = self.P * self.error + \
-                            self.I * self.totalError + \
-                            self.D * (self.error - self.prevError) + \
-                            self.calculateFeedForward()
+            else:
+                if I != 0:
+                    totalError = self.clamp(
+                        totalError + error,
+                        minimumOutput / I,
+                        maximumOutput / I)
+
+                result = P * error + I * totalError + D * (error - prevError) + \
+                    feedForward
                             
-                self.prevError = self.error
-
-                self.result = self.clamp(self.result, self.minimumOutput, self.maximumOutput)
+            result = self.clamp(result, minimumOutput, maximumOutput)
 
-                pidOutput = self.pidOutput
-                result = self.result
-                
-                self.buf.append(self.error)
+            with self.pidWriteMutex:
+                with self.mutex:
+                    enabled = self.enabled
```

Comment:
Need eyes on this bit. It looks to me like this is equivalent to what's in upstream, but not sure.
It also looks like upstream could have just called isEnabled

Reply:
Yeah, this interpretation looks correct to me.